From 23e4a78ecbaf6a15dbf80346b96874dc71f3aef5 Mon Sep 17 00:00:00 2001
From: Dzmitry Malyshau <dmalyshau@mozilla.com>
Date: Thu, 5 Dec 2019 13:02:14 -0500
Subject: [PATCH] Restrict the slow texture upload workaround to 128bit
 formats. (#21)

The workaround is slow for having a CPU-visible mirror of the texture, but that mirror is not properly recycled with regards to GPU usage, thus causing a CPU stall on update.

Original reports were related to WebRender vertex texture updates that are RGBA32F and RGBA32U formats. Limiting the workaround to these formats would allow the affected platforms to upload regular texture data faster without stalls.

Co-authored-by: Erich Gubler <erichdongubler@gmail.com>
---
 .../platform/autogen/FeaturesD3D_autogen.h    |  6 +++++
 include/platform/d3d_features.json            |  7 ++++++
 src/libANGLE/renderer/d3d/TextureD3D.cpp      | 23 ++++++++++++++++---
 src/libANGLE/renderer/d3d/TextureD3D.h        |  1 +
 .../renderer/d3d/d3d11/renderer11_utils.cpp   |  6 +++--
 util/autogen/angle_features_autogen.cpp       |  1 +
 util/autogen/angle_features_autogen.h         |  1 +
 7 files changed, 40 insertions(+), 5 deletions(-)

diff --git a/include/platform/autogen/FeaturesD3D_autogen.h b/include/platform/autogen/FeaturesD3D_autogen.h
index 8bd269c37..5d38b6eb0 100644
--- a/include/platform/autogen/FeaturesD3D_autogen.h
+++ b/include/platform/autogen/FeaturesD3D_autogen.h
@@ -32,6 +32,12 @@ struct FeaturesD3D : FeatureSetBase
         &members,
     };
 
+    FeatureInfo setDataFasterThanImageUploadOn128bitFormats = {
+        "setDataFasterThanImageUploadOn128bitFormats",
+        FeatureCategory::D3DWorkarounds,
+        &members,
+    };
+
     FeatureInfo depthStencilBlitExtraCopy = {
         "depthStencilBlitExtraCopy",
         FeatureCategory::D3DWorkarounds,
diff --git a/include/platform/d3d_features.json b/include/platform/d3d_features.json
index c718cedff..09972897e 100644
--- a/include/platform/d3d_features.json
+++ b/include/platform/d3d_features.json
@@ -21,6 +21,13 @@
                 "Set data faster than image upload"
             ]
         },
+        {
+            "name": "set_data_faster_than_image_upload_on_128bit_formats",
+            "category": "Workarounds",
+            "description": [
+                "Set data faster than image upload on 128bit formats"
+            ]
+        },
         {
             "name": "depth_stencil_blit_extra_copy",
             "category": "Workarounds",
diff --git a/src/libANGLE/renderer/d3d/TextureD3D.cpp b/src/libANGLE/renderer/d3d/TextureD3D.cpp
index 842388878..932607fcc 100644
--- a/src/libANGLE/renderer/d3d/TextureD3D.cpp
+++ b/src/libANGLE/renderer/d3d/TextureD3D.cpp
@@ -366,6 +366,23 @@ angle::Result TextureD3D::setStorageExternalMemory(const gl::Context *context,
     return angle::Result::Continue;
 }
 
+bool TextureD3D::couldUseSetData() const
+{
+    if (!mRenderer->getFeatures().setDataFasterThanImageUpload.enabled)
+    {
+        return false;
+    }
+
+    if (!mRenderer->getFeatures().setDataFasterThanImageUploadOn128bitFormats.enabled)
+    {
+        gl::InternalFormat internalFormat =
+            gl::GetSizedInternalFormatInfo(getBaseLevelInternalFormat());
+        return internalFormat.pixelBytes < 16;
+    }
+
+    return true;
+}
+
 bool TextureD3D::shouldUseSetData(const gl::ImageIndex &index, const ImageD3D *image) const
 {
     if (!isValidIndex(index) || !isImageComplete(index))
@@ -373,7 +390,7 @@ bool TextureD3D::shouldUseSetData(const gl::ImageIndex &index, const ImageD3D *i
         return false;
     }
 
-    if (!mRenderer->getFeatures().setDataFasterThanImageUpload.enabled)
+    if (!couldUseSetData())
     {
         return false;
     }
@@ -775,13 +792,13 @@ angle::Result TextureD3D::generateMipmapUsingImages(const gl::Context *context,
         canCreateRenderTargetForImage(getImageIndex(baseLevel, 0)) &&
         mRenderer->getRendererClass() == RENDERER_D3D11)
     {
-        if (!mRenderer->getFeatures().setDataFasterThanImageUpload.enabled)
+        if (!couldUseSetData())
         {
             ANGLE_TRY(updateStorage(context));
         }
         ANGLE_TRY(ensureRenderTarget(context));
     }
-    else if (mRenderer->getFeatures().setDataFasterThanImageUpload.enabled && mTexStorage)
+    else if (couldUseSetData() && mTexStorage)
     {
         // When making mipmaps with the setData workaround enabled, the texture storage has
         // the image data already. For non-render-target storage, we have to pull it out into
diff --git a/src/libANGLE/renderer/d3d/TextureD3D.h b/src/libANGLE/renderer/d3d/TextureD3D.h
index 3117f08c5..e100ad0e6 100644
--- a/src/libANGLE/renderer/d3d/TextureD3D.h
+++ b/src/libANGLE/renderer/d3d/TextureD3D.h
@@ -259,6 +259,7 @@ class TextureD3D : public TextureImpl, public angle::ObserverInterface
 
     virtual angle::Result updateStorage(const gl::Context *context) = 0;
 
+    bool couldUseSetData() const;
     bool shouldUseSetData(const gl::ImageIndex &index, const ImageD3D *image) const;
     bool isImageSubresourceMatchingStorage(const gl::ImageIndex &index,
                                            const ImageD3D *image,
diff --git a/src/libANGLE/renderer/d3d/d3d11/renderer11_utils.cpp b/src/libANGLE/renderer/d3d/d3d11/renderer11_utils.cpp
index a0feee978..6615f095c 100644
--- a/src/libANGLE/renderer/d3d/d3d11/renderer11_utils.cpp
+++ b/src/libANGLE/renderer/d3d/d3d11/renderer11_utils.cpp
@@ -2188,8 +2188,10 @@ void InitializeFeatures(const Renderer11DeviceCaps &deviceCaps,
                             isIntel && capsVersion >= angle::VersionTriple(16, 0, 0) &&
                                 capsVersion < angle::VersionTriple(16, 4815, 0));
 
-    // Haswell/Ivybridge drivers occasionally corrupt (small?) (vertex?) texture data uploads.
-    ANGLE_FEATURE_CONDITION(features, setDataFasterThanImageUpload,
+    // Haswell drivers occasionally corrupt (small?) (vertex?) texture data uploads for 128bit
+    // formats.
+    ANGLE_FEATURE_CONDITION(features, setDataFasterThanImageUpload, true);
+    ANGLE_FEATURE_CONDITION(features, setDataFasterThanImageUploadOn128bitFormats,
                             !(isIvyBridge || isBroadwell || isHaswell));
 
     ANGLE_FEATURE_CONDITION(features, disableB5G6R5Support,
diff --git a/util/autogen/angle_features_autogen.cpp b/util/autogen/angle_features_autogen.cpp
index bd0a23f86..946509568 100644
--- a/util/autogen/angle_features_autogen.cpp
+++ b/util/autogen/angle_features_autogen.cpp
@@ -335,6 +335,7 @@ constexpr PackedEnumMap<Feature, const char *> kFeatureNames = {{
     {Feature::ScalarizeVecAndMatConstructorArgs, "scalarizeVecAndMatConstructorArgs"},
     {Feature::SelectViewInGeometryShader, "selectViewInGeometryShader"},
     {Feature::SetDataFasterThanImageUpload, "setDataFasterThanImageUpload"},
+    {Feature::SetDataFasterThanImageUploadOn128bitFormats, "setDataFasterThanImageUploadOn128bitFormats"},
     {Feature::SetNeedInitOnInvalidation, "setNeedInitOnInvalidation"},
     {Feature::SetPrimitiveRestartFixedIndexForDrawArrays, "setPrimitiveRestartFixedIndexForDrawArrays"},
     {Feature::SetZeroLevelBeforeGenerateMipmap, "setZeroLevelBeforeGenerateMipmap"},
diff --git a/util/autogen/angle_features_autogen.h b/util/autogen/angle_features_autogen.h
index 6d6b17a8e..147bd30bd 100644
--- a/util/autogen/angle_features_autogen.h
+++ b/util/autogen/angle_features_autogen.h
@@ -335,6 +335,7 @@ enum class Feature
     ScalarizeVecAndMatConstructorArgs,
     SelectViewInGeometryShader,
     SetDataFasterThanImageUpload,
+    SetDataFasterThanImageUploadOn128bitFormats,
     SetNeedInitOnInvalidation,
     SetPrimitiveRestartFixedIndexForDrawArrays,
     SetZeroLevelBeforeGenerateMipmap,
-- 
2.56.0

